ports/mimxrt/machine_sdcard.c

```
@@ -0,0 +1,228 @@
+    CLOCK_SetDiv(kCLOCK_Usdhc1Div, 1U);  // USDHC_input_clk = PFD0_clk / 2 = 198MHZ
+    CLOCK_SetMux(kCLOCK_Usdhc1Mux, 1U);  // Select PFD0 as clock input for USDHC
+    #endif
+
+    #if defined MICROPY_USDHC2 && USDHC2_AVAIL
+    // Configure USDHC clock source and divider
+    CLOCK_SetDiv(kCLOCK_Usdhc2Div, 1U);  // USDHC_input_clk = PFD0_clk / 2 = 198MHZ
+    CLOCK_SetMux(kCLOCK_Usdhc2Mux, 1U);  // Select PFD0 as clock input for USDHC
+    #endif
+
+    // Initialize USDHC
+    const usdhc_config_t config = {
+        .endianMode = kUSDHC_EndianModeLittle,
+        .dataTimeout = 0xFU,
+        .readWatermarkLevel = 128U,
+        .writeWatermarkLevel = 128U,
+    };
+    USDHC_Init(self->usdhc_inst, &config);
+
+    sdcard_init(self, 198000000UL);  // Initialize SDCard Host with 198MHz base clock
+    sdcard_init_pins(self);
+
+    self->initialized = false;
+
+    ticks_delay_us64(2ULL * 1000ULL);  // Wait 2ms to allow USDHC signals to settle/debounce
+
+    return sdcard_detect(self);
+}
+
+STATIC mp_obj_t sdcard_obj_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    mp_map_t kw_args;
+    mp_map_init_fixed_table(&kw_args, n_kw, all_args + n_args);
+
+    // Parse args
+    mp_arg_val_t args[MP_ARRAY_SIZE(sdcard_init_allowed_args)];
+    mp_arg_parse_all(n_args, all_args, &kw_args, MP_ARRAY_SIZE(sdcard_init_allowed_args), sdcard_init_allowed_args, args);
+
+    // Extract arguments
+    mp_int_t sdcard_id = args[SDCARD_INIT_ARG_ID].u_int;
+
+    if (!(1 <= sdcard_id && sdcard_id <= MP_ARRAY_SIZE(mimxrt_sdcard_objs))) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "SDCard(%d) doesn't exist", sdcard_id));
+    }
+
+    mimxrt_sdcard_obj_t *self = &mimxrt_sdcard_objs[(sdcard_id - 1)];
+
+    // Initialize SDCard Host
+    if (machine_sdcard_init_helper(self, args)) {
+        return MP_OBJ_FROM_PTR(self);
+    } else {
+        return mp_const_none;
```

It is highly unconventional to return `None` from a constructor like this.  There is no corresponding way to do this in pure-Python (unless you return `None` from `__new__`).  The user is always expecting a constructor to return an actual instance of the type, or raise an exception.  So this should really raise an exception on error.

---

Ok, I got your point. You are right, that makes not a lot of sense here. I believe it might be better if I rework it such that the constructor always returns an instance. This would be in line with the other ports as far as I could see. Checking for a card can be done once you try to access it. What do you think about that?

---

> Checking for a card can be done once you try to access it.

Yes this would be better.

In stm32, there is an `SDCard.present()` method, which just checks if the SDCard switch is active or not.  In esp32 there is `SDCard.info()` which I think will raise an exception if the card is not there.

---

Ok I will change it today. Thanks for your review again. Highly appreciated.

---

fixed.